shared_model/interfaces/iroha_internal/transaction_sequence.cpp

```diff
@@ -25,21 +25,54 @@ namespace shared_model {
 
     template iroha::expected::Result<TransactionSequence, std::string>
     TransactionSequence::createTransactionSequence(
-        const types::TransactionsForwardCollectionType &transactions,
+        const types::SharedTxsCollectionType &transactions,
         const validation::TransactionsCollectionValidator<
             validation::TransactionValidator<
                 validation::FieldValidator,
                 validation::CommandValidatorVisitor<
                     validation::FieldValidator>>> &validator);
 
-    types::TransactionsForwardCollectionType
-    TransactionSequence::transactions() {
+    types::SharedTxsCollectionType TransactionSequence::transactions() const {
       return transactions_;
     }
 
     TransactionSequence::TransactionSequence(
-        const types::TransactionsForwardCollectionType &transactions)
+        const types::SharedTxsCollectionType &transactions)
         : transactions_(transactions) {}
 
+    types::BatchesType TransactionSequence::batches() const {
+      if (batches_) {
+        return batches_.value();
+      }
+
+      std::unordered_map<std::string, std::vector<std::shared_ptr<Transaction>>>
+          extracted_batches;
+      std::vector<types::SharedTxsCollectionType> batches;
+      for (const auto &tx : transactions_) {
+        if (auto meta = tx->batchMeta()) {
+          auto hashes = meta.get()->transactionHashes();
+          auto batch_hash = this->calculateBatchHash(hashes);
+          extracted_batches[batch_hash].push_back(tx);
+        } else {
+          batches.emplace_back(std::vector<std::shared_ptr<Transaction>>{tx});
+        }
+      }
+      for (auto it : extracted_batches) {
+        batches.emplace_back(it.second);
+      }
+
+      batches_.emplace(batches);
+      return batches;
+    }
```

Comment:
Why not Lazy?

Reply:
I've tried. But it resulted into runtime error. I don't know why